src/emitters/distant.cpp

```diff
@@ -0,0 +1,157 @@
+#include <mitsuba/core/warp.h>
+#include <mitsuba/render/emitter.h>
+#include <mitsuba/render/scene.h>
+#include <mitsuba/render/texture.h>
+
+NAMESPACE_BEGIN(mitsuba)
+
+/**!
+
+.. _emitter-distant:
+
+Distant directional emitter (:monosp:`distant`)
+-----------------------------------------------
+
+.. pluginparameters::
+
+    * - irradiance
+      - |spectrum|
+      - Spectral irradiance, i.e. amount of spectral power per unit area 
+        received by a hypothetical surface normal to the specified direction.
+    * - to_world
+      - |transform|
+      - Emitter-to-world transformation matrix.
+    * - direction
+      - |vector|
+      - Alternative (and exclusive) to `to_world`. Direction towards which the
+        emitter is radiating in world coordinates.
+
+This emitter plugin implements a distant directional source which radiates a
+specified power per unit area along a fixed direction. By default, the emitter
+radiates in the direction of the positive Z axis, i.e. :math:`(0, 0, 1)`.
+
+*/
+
+template <typename Float, typename Spectrum>
+class DistantEmitter final : public Emitter<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(Emitter, m_flags, m_world_transform)
+    MTS_IMPORT_TYPES(Scene, Texture)
+
+    DistantEmitter(const Properties &props) : Base(props) {
+        /* Until `set_scene` is called, we have no information
+           about the scene and default to the unit bounding sphere. */
+        m_bsphere = ScalarBoundingSphere3f(ScalarPoint3f(0.f), 1.f);
+
+        if (props.has_property("direction")) {
+            if (props.has_property("to_world"))
+                Throw("Only one of the parameters 'direction' and 'to_world' "
+                      "can be specified at the same time!'");
+
+            ScalarVector3f direction(normalize(props.vector3f("direction")));
+            auto [up, unused] = coordinate_system(direction);
+
+            m_world_transform =
+                new AnimatedTransform(ScalarTransform4f::look_at(
+                    ScalarPoint3f(0.0f), ScalarPoint3f(direction), up));
+        }
+
+        m_flags      = EmitterFlags::Infinite | EmitterFlags::DeltaDirection;
+        m_irradiance = props.texture<Texture>("irradiance", Texture::D65(1.f));
+    }
+
+    void set_scene(const Scene *scene) override {
+        m_bsphere = scene->bbox().bounding_sphere();
+        m_bsphere.radius =
+            max(math::RayEpsilon<Float>,
+                m_bsphere.radius * (1.f + math::RayEpsilon<Float>) );
+    }
+
+    Spectrum eval(const SurfaceInteraction3f &si, Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointEvaluate, active);
+        Vector3f v = m_world_transform->eval(si.time, active)
+                         .inverse()
+                         .transform_affine(-si.wi);
+        return (allclose(v, Vector3f{ 0.f, 0.f, 1.f }))
+                   ? unpolarized<Spectrum>(m_irradiance->eval(si, active))
+                   : unpolarized<Spectrum>(0.f);
```

Comment:
Here you should use a `select` statement as `v` might be a array of vectors (e.g. GPU variant).

Reply:
Also, this light having a `DeltaDirection`, should `eval` always return `0.f`?

Reply:
Regarding the use of `select`, I'm still missing an equivalent to `allclose` which would only operate on each packet channel. This is not working as intended:
```suggestion
        return select(allclose(v, Vector3f{ 0.f, 0.f, 1.f }),
                      unpolarized<Spectrum>(m_irradiance->eval(si, active)),
                      unpolarized<Spectrum>(0.f));
```

As for the return value of `eval`, came up with this behaviour by reading the Mitsuba 1 implementation; but I honestly don't know what all the consequences could be, maybe I could miss something?

Reply:
Mitsuba 2 doesn't have this notion of `measure`, I would say we are always in `ESolidAngle` therefore `eval` and `pdf` should always return `0.f`

Reply:
How about that? I basically did what is in `allclose`, without the nested horizontal operation.

```suggestion
        constexpr float rtol = 1e-5f;
        constexpr float atol = 1e-8f;

        auto mask = all(abs(ScalarVector3f{ 0.f, 0.f, 1.f } - v) <=
                        abs(v) * rtol + atol);

        return select(mask, 
                      unpolarized<Spectrum>(m_radiance->eval(si, active)),
                      unpolarized<Spectrum>(0.f));
```

Reply:
I would still argue that this is wrong as the probability of sampling the peak of a delta distribution is zero. You can have a look at `dielectric.cpp` which is a `Delta` BSDF (same idea). There `eval` and `pdf` always return `0.f`, so it should be the same here.

Reply:
Ok, I fixed that.